src/Common/src/Interop/Windows/winhttp/Interop.SafeWinHttpHandle.cs

```diff
@@ -30,6 +30,31 @@ public static void DisposeAndClearHandle(ref SafeWinHttpHandle winHttpHandler)
                 }
             }
 
+            // The SafeHandle base class is inconsistent w.r.t. how Dispose() and DangerousRelease()
+            // are implemented. Both methods can cause the ref count on the SafeHandle to be decremented.
+            // And when the ref count reaches zero, ReleaseHandle() is called and the handle is marked
+            // as 'closed'. However, there is a separate internal state bit to track Dispose() calls.
+            // Calling Dispose() more than once is a no-op.  However, calling DangerousRelease() and then
+            // Dispose() will cause ObjectDisposedException to be thrown if the previous DangerousRelease()
+            // caused the handle to be closed. This seems inconsistent especially since both methods can
+            // decrement the ref count and cause ReleaseHandle() to be called.
+            //
+            // The SafeWinHttpHandle class requires special lifetime management. Multiple classes such as
+            // WinHttpHandler, WinHttpRequestStream, and WinHttpResponseStream interact together with the
+            // same WinHttp handles for a single request/response operation. In addition, there is a parent-child
+            // hierarchy among the handles which causes additional ref counts to be added to keep the handles
+            // valid. Due to these requirements, the use pattern of SafeWinHttpHandle typically uses DangerousAddRef()
+            // and DangerousRelease() to help manage ref counts. Thus, to work around the limitations of the
+            // current SafeHandle implementation, it is important that Dispose() be a no-op if the handle has
+            // already been closed due to final release caused by DangerousRelease().
```

Comment:
I'll read through this more to understand what's being done, but for a parent/child relationship of safe handles, the pattern I've seen is to have the child AddRef its parent on construction and ReleaseRef its parent when the child is disposed/finalized.  That way the parent wont' have its ReleaseHandle called until all children that depend on that same handle have gone away, at which point the ref count on the parent will be back at 0 and finalization will cause the handle to be released (or a delayed previous Dispose that didn't previously take effect because there was a positive ref count).